compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2358,13 +2358,20 @@ impl<'a> Parser<'a> {
     /// Parses a `let $pat = $expr` pseudo-expression.
     /// The `let` token has already been eaten.
     fn parse_let_expr(&mut self, attrs: AttrVec) -> PResult<'a, P<Expr>> {
-        if !self.let_expr_allowed {
-            self.struct_span_err(
-                self.prev_token.span,
-                "expected expression, found `let` statement",
-            )
-            .emit();
+        // Catches things like `if let Some(_) = _opt && [1, 2, 3][let _ = ()] = 1`
+        let is_in_a_let_chains_context_but_nested_in_other_expr = self.let_expr_allowed
+            && !matches!(
+                self.prev_token.kind,
+                TokenKind::AndAnd
+                    | TokenKind::CloseDelim(Delimiter::Brace)
+                    | TokenKind::Ident(kw::If, _)
+                    | TokenKind::Ident(kw::While, _)
+            );
```

Comment:
As you already mention, this is a bit of a hack, but *seems* safe enough (although I feel like it might also not trigger with things like `let _ = &&let Some(x) = Some(42);` where the intent might be `&&bool`, can we test that?).

Having said that, I'm wondering if you explored adding a `visit_pat` function to https://github.com/rust-lang/rust/blob/master/compiler/rustc_ast_passes/src/ast_validation.rs to catch the nested pattern cases (like parens around the `let` expression).

It feels like cases like 
https://github.com/rust-lang/rust/pull/98633/files#diff-211c1d6e2f4ff5f2af12400a49c018c5ecebd9e0ec9e8a15049a9e55521692bbR366-R370
in line 368 (`[][let () = ()]`) should be able to be caught by the current code (but that test has that expression `cfg`d away, which is likely why it wasn't triggering?).

If the reason for this code is to trigger even in the face of `cfg`d code, and modifying the visitor to be more exhaustive isn't possible for good reason, then I'm ok with the code as is. It isn't great, but worst case scenario this is overly restrictive, which is better than the alternative.